### PR TITLE
feat(common/models):  context tracking of accepted Suggestions

### PR DESCRIPTION
**Pull request by jahorton**
This PR adds internal tracking of user manipulations to the context state.  In particular, whenever a user 'accepts' a `Suggestion`, this will be noted within the `ContextTracker` to provide enhanced information to subsequent prediction efforts.  (All such tracking is worker-internal and only for the user's local use.)  This was previously part of #3647, but the changes were distinct enough to merit splitting it in twain - hence this now being reviewable separately.

- Tracked context now notes any accepted suggestions and the resulting token text, maintaining both the original and replaced versions of each token.
- Fixes some related internal context tracking bugs that hadn't yet been caught due to lack of external impact.  (Arguably, it may be more of an "incomplete implementation" scenario.)
- 'Reversions' will still seem a bit broken for now - part of that whole "incomplete implementation" bit.
    - That said, once work is complete, these changes will add extra polish beyond what was available in prior versions of Keyman's predictive text.

Related upcoming features:
- Whenever a user chooses to "revert" a `Suggestion`, the _full_ original context will be restored.
    - The system will remember all calculations for the context, thus allowing the original keystroke probabilities to drive  predictions when extending typed context from a reversion!
    - At the time of reversion, we'll be able to immediately return the original set of predictions, rather than recomputing them from scratch.
- If handled correctly, KMW should be able to 'revert' any deadkeys that were originally 'consumed' when accepting Suggestions.
    - Of course, it's quite debatable whether or not we want to actually do this.
- Debatably, we may be getting close to a point where we _**could**_ experiment with some auto-correction behavior.
    - It's far safer when when no information about the prior state is lost _and_ when it can be easily restored.
    - This particular point will be its own PR if and when we choose to pursue it.

All of this tracking data is _also_ designed for further extensions in the far future, beyond 14.0:
- If we ever wish to allow phrase-level suggestions, the tokenized context tracking is designed to facilitate this.
    - Everything I can foresee such a feature needing will be available there.
    - Also handy for catching cases where 'whitespace typos' occur, which 14.0 predictive text assumes to be impossible.

While a bit more test development probably "wouldn't hurt," there is a test now, so I'm comfortable moving this out of draft.

### CODE REVIEW

File: common/models/types/index.d.ts [hunk 1]
```diff
@@ -216,6 +216,14 @@ declare interface Suggestion {
    */
   transformId?: number;
 
+  /**
+   * A unique identifier for the Suggestion itself, not shared with any others -
+   * even for Suggestions sourced from the same Transform. 
+   * 
+   * The lm-layer is responsible for setting this field, not models.
+   */
+  id?: number;
```
Comment (darcywong00) [inline common/models/types/index.d.ts:225]: Is this more of a guid/uuid? Just to avoid confusion with all the other id's we have around (Transform.id?)

Comment (jahorton) [reply common/models/types/index.d.ts:225]: It is not.  This one is rooted on the new `SUGGESTION_ID_SEED` from `ModelCompositor`.

**Review by mcdurdin — COMMENTED** [review]
I think this looks okay but would like to have @eddieantonio's perspective on it.

**Review by eddieantonio — APPROVED** [review]
LGTM

**Pull request merged**